mne/beamformer/_lcmv.py

```diff
@@ -102,52 +133,90 @@ def _apply_lcmv(data, info, tmin, forward, noise_cov, data_cov, reg,
         Ck = np.dot(Wk, Gk)
 
         # Find source orientation maximizing output source power
-        if pick_ori == 'max-power':
-            eig_vals, eig_vecs = linalg.eigh(Ck)
-
-            # Choosing the eigenvector associated with the middle eigenvalue.
-            # The middle and not the minimal eigenvalue is used because MEG is
-            # insensitive to one (radial) of the three dipole orientations and
-            # therefore the smallest eigenvalue reflects mostly noise.
-            for i in range(3):
-                if i != eig_vals.argmax() and i != eig_vals.argmin():
-                    idx_middle = i
-
-            # TODO: The eigenvector associated with the smallest eigenvalue
-            # should probably be used when using combined EEG and MEG data
-            max_ori = eig_vecs[:, idx_middle]
 
+        # special case 1: weight normalization and orientation selection
+        if weight_norm is not None and pick_ori == 'max-power':
+            # finding optimal orientation for NAI and unit-noise gain
+            # based on [2]_, Eq. 4.47
+            tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+            eig_vals, eig_vecs = linalg.eig(np.dot(linalg.pinv(tmp),
+                                                   np.dot(Wk, Gk)))
+            idx_max = eig_vals.argmax()
+            max_ori = eig_vecs[:, idx_max]
             Wk[:] = np.dot(max_ori, Wk)
-            Ck = np.dot(max_ori, np.dot(Ck, max_ori))
+            Gk = np.dot(Gk, max_ori)
+
+            if weight_norm == 'nai':
+                # compute spatial filter for NAI
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(noise * tmp)
+                Wk /= denom
+            elif weight_norm == 'unit-noise-gain':
+                # compute spatial filter for unit-noise gain
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(tmp)
+                Wk /= denom
+
             is_free_ori = False
 
-        if is_free_ori:
-            # Free source orientation
-            Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+        # special case 2: NAI without orientation selection
+        elif weight_norm == 'nai' and pick_ori != 'max-power':
+            raise ValueError('Weight normalization with Neural activity index '
+                             'is not implemented yet with free or normal '
+                             'orientation.')
+
+        # all other combinations are handled here:
         else:
-            # Fixed source orientation
-            Wk /= Ck
+            if pick_ori == 'max-power':
```

Comment:
this branch will be called if a user does

pick_ori = 'max-power' and weight_norm = None

this option has never been allowed and this code is certainly broken. We should remove it as part of the bug fix.